In [ ]:
from google.colab import drive
drive.mount('/content/drive')
import pickle
# Load the pickle file for word2vec model
with open('/content/drive/MyDrive/Kaggle/word2vec_model.pkl', 'rb') as f:
    word2vec_model = pickle.load(f)

# Load the pickle file for dataset
with open('/content/drive/MyDrive/Kaggle/data_after_word2vec.pkl', 'rb') as f:
    data = pickle.load(f)

data = data.drop(columns=['Unnamed: 0.3', 'Unnamed: 0.2', 'Unnamed: 0.1', 'Unnamed: 0'])
data['cleaned_reviews'] = data['cleaned_reviews'].apply(lambda tokens: ' '.join(tokens))
data.head()

Mounted at /content/drive


url  name  online_order  \
0  https://www.zomato.com/bangalore/jalsa-banasha...  2933             1   
1  https://www.zomato.com/bangalore/jalsa-banasha...  2933             1   
2  https://www.zomato.com/bangalore/jalsa-banasha...  2933             1   
3  https://www.zomato.com/bangalore/jalsa-banasha...  2933             1   
4  https://www.zomato.com/bangalore/jalsa-banasha...  2933             1   

   book_table      rate     votes  location  rest_type  cuisines      cost  \
0           1  0.720882  0.611201         1         24      1972  0.557298   
1           1  0.720882  0.611201         1         24      1972  0.557298   
2           1  0.720882  0.611201         1         24      1972  0.557298   
3           1  0.720882  0.611201         1         24      1972  0.557298   
4           1  0.720882  0.611201         1         24      1972  0.557298   

                                        reviews_list  type  city  \
0  ('Rated 4.0', 'RATED\n  A beautiful place to d...     0     1   
1  ('Rated 4.0', 'RATED\n  I was here for dinner ...     0     1   
2  ('Rated 2.0', 'RATED\n  Its a restaurant near ...     0     1   
3  ('Rated 4.0', 'RATED\n  We went here on a week...     0     1   
4  ('Rated 5.0', 'RATED\n  The best thing about t...     0     1   

   Sentiment_Score  Sentiment_Class  \
0           0.7430                3   
1           0.9486                4   
2           0.7964                3   
3           0.9622                4   
4           0.9461                4   

                               modified_reviews_list  \
0  A beautiful place to dine in.The interiors tak...   
1  I was here for dinner with my family on a week...   
2  Its a restaurant near to Banashankari BDA. Me ...   
3  We went here on a weekend and one of us had th...   
4  The best thing about the place is itÃ\x83\x83Ã...   

                                     cleaned_reviews  
0  beautiful place dine interior take back mughal...  
1  dinner family weekday restaurant completely em...  
2  restaurant near banashankari bda along office ...  
3  go weekend one u buffet two u take ala carte f...  
4  best thing place ambiance second best thing yu...

In [ ]:
vocab_size = len(word2vec_model.wv.key_to_index)
print("Vocabulary Size:", vocab_size)

Vocabulary Size: 56803


In [ ]:
data.columns

Index(['url', 'name', 'online_order', 'book_table', 'rate', 'votes',
       'location', 'rest_type', 'cuisines', 'cost', 'reviews_list', 'type',
       'city', 'Sentiment_Score', 'Sentiment_Class', 'modified_reviews_list',
       'cleaned_reviews'],
      dtype='object')

In [ ]:
#adding the word embedding to the dataset and removing unnecessary columns
import numpy as np
from sklearn.preprocessing import StandardScaler

data.drop(columns=['votes', 'reviews_list', 'modified_reviews_list'], inplace=True)
data.rename(columns={'type': 'dining_type','cost': 'cost(for 2)'}, inplace=True)

scaler = StandardScaler()
columns_to_scale = ['cuisines', 'rest_type']
data[columns_to_scale] = scaler.fit_transform(data[columns_to_scale])
import gensim
def text_to_word_embeddings(text, word2vec_model):
    word_embeddings = []
    for word in text.split():
        if word in word2vec_model.wv:
            word_embeddings.append(word2vec_model.wv[word])
    return np.array(word_embeddings)

data['word_embeddings_of_reviews'] = data['cleaned_reviews'].apply(text_to_word_embeddings, args=(word2vec_model,))
data.head()

url  name  online_order  \
0  https://www.zomato.com/bangalore/jalsa-banasha...  2933             1   
1  https://www.zomato.com/bangalore/jalsa-banasha...  2933             1   
2  https://www.zomato.com/bangalore/jalsa-banasha...  2933             1   
3  https://www.zomato.com/bangalore/jalsa-banasha...  2933             1   
4  https://www.zomato.com/bangalore/jalsa-banasha...  2933             1   

   book_table      rate  location  rest_type  cuisines  cost(for 2)  \
0           1  0.720882         1  -0.684284    1.1181     0.557298   
1           1  0.720882         1  -0.684284    1.1181     0.557298   
2           1  0.720882         1  -0.684284    1.1181     0.557298   
3           1  0.720882         1  -0.684284    1.1181     0.557298   
4           1  0.720882         1  -0.684284    1.1181     0.557298   

   dining_type  city  Sentiment_Score  Sentiment_Class  \
0            0     1           0.7430                3   
1            0     1           0.9486                4   
2            0     1           0.7964                3   
3            0     1           0.9622                4   
4            0     1           0.9461                4   

                                     cleaned_reviews  \
0  beautiful place dine interior take back mughal...   
1  dinner family weekday restaurant completely em...   
2  restaurant near banashankari bda along office ...   
3  go weekend one u buffet two u take ala carte f...   
4  best thing place ambiance second best thing yu...   

                          word_embeddings_of_reviews  
0  [[2.0981128, 1.815063, -6.8484993, 1.6597033, ...  
1  [[0.77261764, 3.2976336, 1.0144088, -5.099168,...  
2  [[4.081995, 4.763423, -2.1862926, 3.2217922, -...  
3  [[2.0472045, 0.7242372, 3.7346416, -0.51295847...  
4  [[1.6863357, 7.823443, -0.32242757, -0.3208254...

In [ ]:
data['location'].nunique()

90

In [ ]:
# location = 22 is the BTM neighbourhood  Byrasandra, Tavarekere and Madiwala Layout
data['location'].value_counts().sort_values(ascending=False)

44    132029
0      84821
27     73570
22     64408
55     56257
       ...  
81        29
38        27
23        17
88        15
65         9
Name: location, Length: 90, dtype: int64

In [ ]:
# Assuming data is a pandas DataFrame
urls_for_location_12 = data.loc[data['location'] == 22, 'url']
urls_for_location_12[54161]

'https://www.zomato.com/bangalore/navya-foods-hsr-bangalore?context=eyJzZSI6eyJlIjpbIjUzODM0IiwiMTg1MzE2NTAiLCIxODYxMzY2NyIsIjE4NjgwOTAxIiwxODg4OTM0NywiMTg1NTkyMTEiLCIxODY0MjkzMyIsIjUwODkxIiwiMTg3Nzc0MjgiLCI1NzI5MiIsIjU3Nzg1IiwiMTgzNTE0NDAiLCI1MTcwOSIsIjE4NTAwMDIwIiwiMTg3MzYwNjYiLCIxODg5Mzk2MyIsIjE4MzM5ODc3IiwiMTg2MzYzMTAiLCIxODc4MjQ2OSIsIjE4OTExMDE0IiwiMTg4OTQ5MzUiLCI1Nzg2NyJdLCJ0IjoiRGVsaXZlcnkgUmVzdGF1cmFudHMgaW4gQmFzYXZhbmFndWRpIn19'

In [ ]:
new_df = data[data['location'] == 22]

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
# columns_to_scale = ['dining_type', 'city']
new_df[['dining_type']] = scaler.fit_transform(new_df[['dining_type']])
new_df.head()

<ipython-input-9-7155f57a08fe>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df[['dining_type']] = scaler.fit_transform(new_df[['dining_type']])


url  name  online_order  \
54161  https://www.zomato.com/bangalore/navya-foods-h...  4124             1   
69129  https://www.zomato.com/bangalore/the-paratha-c...  6375             1   
69130  https://www.zomato.com/bangalore/the-paratha-c...  6375             1   
69131  https://www.zomato.com/bangalore/the-paratha-c...  6375             1   
69132  https://www.zomato.com/bangalore/the-paratha-c...  6375             1   

       book_table      rate  location  rest_type  cuisines  cost(for 2)  \
54161           0  0.286423        22   1.278466  1.001651    -0.582052   
69129           0  0.845013        22  -0.203611  0.647824    -0.354182   
69130           0  0.845013        22  -0.203611  0.647824    -0.354182   
69131           0  0.845013        22  -0.203611  0.647824    -0.354182   
69132           0  0.845013        22  -0.203611  0.647824    -0.354182   

       dining_type  city  Sentiment_Score  Sentiment_Class  \
54161    -0.486294     3           0.9002                4   
69129    -0.486294     4           0.7885                3   
69130    -0.486294     4           0.9623                4   
69131    -0.486294     4           0.8297                4   
69132    -0.486294     4           0.9854                4   

                                         cleaned_reviews  \
54161  boneless biriyani good like much homely foods ...   
69129  x ordered place multiple time love food favour...   
69130  new outlet hsr serve best food hsr guess try a...   
69131  place order aloo paratha dal makhani box parat...   
69132  go zomato find place place order aloo paratha ...   

                              word_embeddings_of_reviews  
54161  [[0.72709286, 2.2768533, 3.0795612, -5.3143263...  
69129  [[3.6359832, -0.41793633, 2.6480174, -0.356185...  
69130  [[2.5347857, 0.22424214, 1.9644408, 3.3490353,...  
69131  [[3.2815177, 2.9132953, -1.1734662, 2.2773705,...  
69132  [[2.0472045, 0.7242372, 3.7346416, -0.51295847...

In [ ]:
new_df.shape

(64408, 15)

In [ ]:
import pickle
with open('/content/drive/MyDrive/Kaggle/data location = 22(BTM).pkl', 'wb') as f:
  pickle.dump(new_df, f)



---



In [ ]:
from google.colab import drive
drive.mount('/content/drive')

import pickle
# Load the pickle file for dataset
with open('/content/drive/MyDrive/Kaggle/data location = 22(BTM).pkl', 'rb') as f:
    data_btm = pickle.load(f)

Mounted at /content/drive


In [ ]:
data_btm.shape

(64408, 15)

In [ ]:
data_btm.head()

url  name  online_order  \
54161  https://www.zomato.com/bangalore/navya-foods-h...  4124             1   
69129  https://www.zomato.com/bangalore/the-paratha-c...  6375             1   
69130  https://www.zomato.com/bangalore/the-paratha-c...  6375             1   
69131  https://www.zomato.com/bangalore/the-paratha-c...  6375             1   
69132  https://www.zomato.com/bangalore/the-paratha-c...  6375             1   

       book_table      rate  location  rest_type  cuisines  cost(for 2)  \
54161           0  0.286423        22   1.278466  1.001651    -0.582052   
69129           0  0.845013        22  -0.203611  0.647824    -0.354182   
69130           0  0.845013        22  -0.203611  0.647824    -0.354182   
69131           0  0.845013        22  -0.203611  0.647824    -0.354182   
69132           0  0.845013        22  -0.203611  0.647824    -0.354182   

       dining_type  city  Sentiment_Score  Sentiment_Class  \
54161    -0.486294     3           0.9002                4   
69129    -0.486294     4           0.7885                3   
69130    -0.486294     4           0.9623                4   
69131    -0.486294     4           0.8297                4   
69132    -0.486294     4           0.9854                4   

                                         cleaned_reviews  \
54161  boneless biriyani good like much homely foods ...   
69129  x ordered place multiple time love food favour...   
69130  new outlet hsr serve best food hsr guess try a...   
69131  place order aloo paratha dal makhani box parat...   
69132  go zomato find place place order aloo paratha ...   

                              word_embeddings_of_reviews  
54161  [[0.72709286, 2.2768533, 3.0795612, -5.3143263...  
69129  [[3.6359832, -0.41793633, 2.6480174, -0.356185...  
69130  [[2.5347857, 0.22424214, 1.9644408, 3.3490353,...  
69131  [[3.2815177, 2.9132953, -1.1734662, 2.2773705,...  
69132  [[2.0472045, 0.7242372, 3.7346416, -0.51295847...

# SVM
---
(WITHOUT WORD EMBEDDINGS)



In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, recall_score
from sklearn.svm import SVC
data_btm['Sentiment_Class'] = data_btm['Sentiment_Class'].astype(int)

x = data_btm[['online_order', 'book_table', 'rate', 'rest_type', 'cuisines', 'cost(for 2)', 'dining_type' ]]
y = data_btm['Sentiment_Class']

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [ ]:
x_train.shape

(51526, 7)

In [ ]:
data_btm['Sentiment_Class'].value_counts()

4    31409
3    15565
2    10030
1     6019
0     1385
Name: Sentiment_Class, dtype: int64

In [ ]:
svm_model = SVC(kernel='poly', C=1.0, decision_function_shape='ovr')
params_grid = [{'kernel': ['rbf'], 'gamma': [1e-3, 1e-4],
                     'C': [1, 10, 100, 1000]},
                    {'kernel': ['linear'], 'C': [1, 10, 100, 1000]}]
svm_model.fit(x_train, y_train)

SVC(kernel='poly')

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

y_pred = svm_model.predict(x_test)
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='weighted')
recall = recall_score(y_test, y_pred, average='weighted')
f1 = f1_score(y_test, y_pred, average='weighted')
conf_matrix = confusion_matrix(y_test, y_pred)

print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)
print("Confusion Matrix:\n", conf_matrix)


Accuracy: 0.5424623505666822
Precision: 0.4549462623328751
Recall: 0.5424623505666822
F1-score: 0.4560381105809704
Confusion Matrix:
 [[   0    1   69    9  196]
 [   0    6  415   90  656]
 [   0    8 1006  218  811]
 [   0   11  606  259 2198]
 [   0   11  438  157 5717]]


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


1. **Accuracy**: The model's predictions are correct for about 54.2% of the instances.

2. **Precision**: The model predicts a positive instance, it is correct about 45.5% of the time.

3. **Recall**: The model correctly identifies about 54.2% of the actual positive instances.

4. **F1-score**: The F1-score is approximately 0.456, reflecting the balance between precision and recall in the model's predictions.

5. **Confusion Matrix**:
  - 0 instances were correctly classified as Class 1 (True Negatives).
  - 1 instance was incorrectly classified as Class 2 (False Positives).
  - 6 instances were correctly classified as Class 2 (True Positives).
  - 1006 instances were correctly classified as Class 3 (True Positives).
  - 259 instances were correctly classified as Class 4 (True Positives).
  - 5717 instances were correctly classified as Class 5 (True Positives).
  - rest were incorrectly classified

{Class 1: very bad, Class 2: bad, Class 3: neutral, Class 4: good, Class 5: very good}




In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.00      0.00      0.00       275
           1       0.16      0.01      0.01      1167
           2       0.40      0.49      0.44      2043
           3       0.35      0.08      0.14      3074
           4       0.60      0.90      0.72      6323

    accuracy                           0.54     12882
   macro avg       0.30      0.30      0.26     12882
weighted avg       0.45      0.54      0.46     12882



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
import pickle

# Define the file path for saving the model
file_path = '/content/drive/MyDrive/Kaggle/svm_model_model.pkl'

# Save the model to the file
with open(file_path, 'wb') as file:
    pickle.dump(svm_model, file)

# SVM
---
(WITH WORD EMBEDDINGS)


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, recall_score
from sklearn.svm import SVC
import numpy as np
data_btm['Sentiment_Class'] = data_btm['Sentiment_Class'].astype(int)
x = data_btm[['online_order', 'book_table', 'rate', 'rest_type', 'cuisines', 'cost(for 2)', 'dining_type', 'word_embeddings_of_reviews' ]]
y = data_btm['Sentiment_Class']
word_embeddings = np.vstack(data_btm['word_embeddings_of_reviews'])
x['word_embeddings_of_reviews'] = word_embeddings
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)
x_train.shape

ValueError: Length of values (1872174) does not match length of index (64408)

In [ ]:
from sklearn.model_selection import GridSearchCV
svm_model = SVC(kernel='poly', C=1.0, decision_function_shape='ovr')
params_grid = [{'kernel': ['rbf'], 'gamma': [1e-3, 1e-4], 'C': [1, 10, 100, 1000]},
               {'kernel': ['linear'], 'C': [1, 10, 100, 1000]}]
grid_search = GridSearchCV(svm_model, params_grid, cv=5)
grid_search.fit(x_train, y_train)

ValueError: 
All the 60 fits failed.
It is very likely that your model is misconfigured.
You can try to debug the error by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
60 fits failed with the following error:
TypeError: only length-1 arrays can be converted to Python scalars

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py", line 192, in fit
    X, y = self._validate_data(
  File "/usr/local/lib/python3.10/dist-packages/sklearn/base.py", line 584, in _validate_data
    X, y = check_X_y(X, y, **check_params)
  File "/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py", line 1106, in check_X_y
    X = check_array(
  File "/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py", line 879, in check_array
    array = _asarray_with_order(array, order=order, dtype=dtype, xp=xp)
  File "/usr/local/lib/python3.10/dist-packages/sklearn/utils/_array_api.py", line 185, in _asarray_with_order
    array = numpy.asarray(array, order=order, dtype=dtype)
  File "/usr/local/lib/python3.10/dist-packages/pandas/core/generic.py", line 2070, in __array__
    return np.asarray(self._values, dtype=dtype)
ValueError: setting an array element with a sequence.


In [ ]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split

# Assuming x and y are your feature matrix and target variable, respectively
# Split the data into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

# Define the SVM model
svm_model = SVC(decision_function_shape='ovr')

# Define the parameter grid for hyperparameter tuning
params_grid = [{'kernel': ['rbf'], 'gamma': [1e-3, 1e-4], 'C': [1, 10, 100, 1000]},
               {'kernel': ['linear'], 'C': [1, 10, 100, 1000]}]

# Perform grid search cross-validation to find the best hyperparameters
grid_search = GridSearchCV(svm_model, params_grid, cv=5)
grid_search.fit(x_train, y_train)

# Print the best hyperparameters found by grid search
print("Best hyperparameters:", grid_search.best_params_)

# Train the SVM model with the best hyperparameters on the entire training set
best_svm_model = SVC(**grid_search.best_params_, decision_function_shape='ovr')
best_svm_model.fit(x_train, y_train)

# Evaluate the model on the test set
y_pred = best_svm_model.predict(x_test)

# Print classification report
print("Classification Report:")
print(classification_report(y_test, y_pred))




---



In [ ]:
data.dtypes

url                            object
name                            int64
online_order                    int64
book_table                      int64
rate                          float64
location                        int64
rest_type                     float64
cuisines                      float64
cost(for 2)                   float64
dining_type                     int64
city                            int64
Sentiment_Score               float64
Sentiment_Class                 int64
cleaned_reviews                object
word_embeddings_of_reviews     object
dtype: object

In [ ]:
numeric_columns = data.select_dtypes(include=['float64', 'int64']).columns

data[numeric_columns] = data[numeric_columns].astype('float32')

In [ ]:
embeddings_array = np.array(data['word_embeddings_of_reviews'].tolist())

# Convert NumPy array to TensorFlow tensor
embeddings_tensor = tf.convert_to_tensor(embeddings_array, dtype=tf.float32)

ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (1316675,) + inhomogeneous part.

In [ ]:
import tensorflow as tf
# from tf.keras.layers.recurrent import LSTM
# X = data[['online_order', 'book_table', 'rate', 'votes', 'rest_type', 'cuisines', 'cost', 'modified_reviews_list']]
y = np.asarray(data['Sentiment_Class']).astype(np.float32)

X = data[['online_order', 'book_table', 'rate', 'location', 'rest_type', 'cuisines', 'cost(for 2)', 'dining_type', 'word_embeddings_of_reviews']]
# y = data['Sentiment_Class']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# # Assuming X_train and y_train are your training features and labels
# # Create a Sequential model
model = tf.keras.Sequential()
vocab_size = 7000
embd_len = 32
max_words = 150
# # Add a Dense layer with one unit (for binary classification) and a linear activation function
# model.add(tf.keras.layers.Dense(units=1, activation='linear'))
model.add(tf.keras.layers.Embedding(vocab_size,
                         embd_len,
                         input_length=max_words))
model.add(tf.keras.layers.LSTM(100, activation='softmax'))
model.add(tf.keras.layers.Dense(1))
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# # Compile the model with a hinge loss function and Adam optimizer
# model.compile(loss='hinge', optimizer='adam', metrics=['accuracy'])

# # Train the model
model.fit(X_train, y_train, epochs=10, batch_size=32, validation_split=0.2)

ValueError: Failed to convert a NumPy array to a Tensor (Unsupported object type float).

In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Embedding, Flatten

# Assuming 'data' is your DataFrame containing the data

# Assuming you have already scaled the numerical features and stored them in the DataFrame
# Assuming 'word_embeddings_of_reviews' contains the word embeddings

# Splitting data into X and y
y = np.asarray(data['Sentiment_Class']).astype(np.float32)
X = data[['online_order', 'book_table', 'rate', 'location', 'rest_type', 'cuisines', 'cost(for 2)', 'dining_type', 'word_embeddings_of_reviews']]

# Splitting the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Define the parameters
vocab_size = len(word2vec_model.wv.key_to_index)
embedding_dim = 100  # You can adjust this as per your requirements
max_length = 100  # You should set this according to your maximum sequence length

# Define the LSTM model
model = Sequential()

# Add an embedding layer
model.add(Embedding(input_dim=vocab_size, output_dim=embedding_dim, input_length=max_length))

# Add LSTM layer
model.add(LSTM(128, return_sequences=True))
model.add(LSTM(64))

# Add a dense layer
model.add(Dense(64, activation='relu'))

# Output layer
model.add(Dense(5, activation='softmax'))  # 5 classes, softmax activation for multiclass classification

# Compile the model
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Train the model
model.fit(X_train, y_train, epochs=10, batch_size=32, validation_split=0.1)

# Evaluate the model
loss, accuracy = model.evaluate(X_test, y_test)
print("Test Accuracy:", accuracy)


NameError: name 'embedding_dim' is not defined

In [ ]:
type(x)
# y_pred = model.predict(X_test)

# import numpy as np
# y_pred = model.predict(X_test)
# y_pred_labels = np.round(y_pred).flatten()
# accuracy_score(y_test, y_pred)

In [ ]:
#reducing the dataset in half
import pandas as pd
import numpy as np

# Calculate the new counts for each value
new_counts = data['name'].value_counts() // 2
new_counts[new_counts == 0] = 1  # Set counts of 1 to 1 (to avoid setting them to 0)

# Create a DataFrame to store the modified data
new_data = pd.DataFrame(columns=data.columns)

# Update the counts in the dataset and randomly select a subset of rows
for name, count in new_counts.items():
    rows = data[data['name'] == name].sample(n=count, random_state=42)  # Randomly select 'count' rows
    new_data = pd.concat([new_data, rows])

# Reset the index of the new DataFrame
new_data.reset_index(drop=True, inplace=True)

# Check the size of the new dataset
print("Original dataset size:", len(data))
print("New dataset size:", len(new_data))

# Check the updated counts
print(new_data['name'].value_counts())

Original dataset size: 1316675
New dataset size: 657323
2454    5942
3757    4573
6591    4346
4418    3814
1553    3202
        ... 
5677       1
4269       1
5676       1
3369       1
1916       1
Name: name, Length: 7006, dtype: int64
